src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/InstrumentAdvice.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// Contains configuration settings advised to be used by metrics consumers when recording measurements for a given <see cref="Instrument{T}"/>.
+    /// </summary>
+    /// <typeparam name="T">Instrument value type.</typeparam>
+    public sealed class InstrumentAdvice<T> where T : struct
+    {
+        private readonly ReadOnlyCollection<T>? _HistogramBucketBoundaries;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="InstrumentAdvice{T}"/> class.
+        /// </summary>
+        public InstrumentAdvice()
+        {
+            Instrument.ValidateTypeParameter<T>();
+        }
+
+        /// <summary>
+        /// Gets the explicit bucket boundaries advised to be used with histogram instruments.
+        /// </summary>
+        /// <remarks>
+        /// Notes:
+        /// <list type="bullet">
+        /// <item>A <see langword="null"/> value means no bucket boundaries have been configured and default values should be used for bucket configuration.</item>
+        /// <item>An empty set of bucket boundaries hints that the histogram by default should NOT contain buckets and should only track count and sum values.</item>
+        /// <item>A set of distinct increasing values for bucket boundaries hints that the histogram should use those for its default bucket configuration.</item>
+        /// </list>
+        /// </remarks>
+        public IReadOnlyList<T>? HistogramBucketBoundaries
+        {
+            get => _HistogramBucketBoundaries;
+            init
```

Comment:
I humbly disagree. As I mentioned before, doing the allocation and copying is not good IMO especially we need to maintain that for future usage. We discussed that in the design review, and I do not see any objection. Also, `init` is already used in the same library too.

Reply:
The code as-is already takes a copy of the data 😄 The issue I think `set` creates is more that it can't be changed after the first publish of a metric. What OTel is going to do is on instrument published take the advice and construct the buckets after which point they are fixed. So having the `set` could just be misleading for users thinking they can set it _whenever_. `init` really conveys the meaning nicely for what this class is doing.

I guess we could have `set` throw based on some internal state tracking if the thing was published? But then what if you hand the same advice class to multiple instruments? 🤔

Reply:
We had a fair bit of internal discussion around init and concluded that yes, we are going to use the new language feature here.

We know that the presence of init makes it less straightforward to consume from projects using older runtime versions such as .NET Framework 4.8. By default those older versions target C# language version 7.3 which doesn't support the init keyword. Our overall goal is to design new feature APIs favoring projects that also use new languages and runtimes if a tradeoff has to be made. For developers who remain on older runtime versions (such as .NET Framework) there are a couple options:
- The most conservative choice would be not to use new features at all if they require a new language version and stick with the features that were available at the time a given runtime was released.
- An alternative that I think many developers would find preferable is to target their project at a newer C# language version when needed. Although old runtimes don't support every new language feature, the compiler does do build-time error checking to prevent accidental usage of new features on runtime versions that are known to be incompatible. The .NET team also validates our features in .NET Framework test apps that invoke the new APIs using the latest C# language version. In practice we see many .NET developers update their C# language version and are very happy with the results.